### PR TITLE
Determine the alias for the columns at the time of execution.

### LINKED ISSUE #10: Add support django 3.0

### PR DESCRIPTION
**Pull request by tim-schilling**
This should close #10

### CODE REVIEW

File: django_cte/meta.py [hunk 1]
```diff
@@ -75,21 +75,8 @@ class CTEColumnRef(Expression):
...
-        ref = self.copy()
-        ref._alias = query.get_initial_alias()
-        return ref
-
-    def relabeled_clone(self, relabels):
-        if self._alias is not None and self._alias in relabels:
-            clone = self.copy()
-            clone._alias = relabels[self._alias]
-            return clone
-        return self
```
Comment (millerdev) [inline django_cte/meta.py]: I'm reluctant to merge this without digging a bit deeper to understand how this method can be safely removed. If it turns out to be unnecessary then this looks like an elegant fix. I see that all tests have passed, which is a good sign. However, I seem to remember some edge cases where `relabeled_clone` was necessary (although I think I would have added tests for them). Can you explain in a bit more detail why this is no longer needed? Out of curiosity, is `relabeled_clone` no longer a thing in Django 3?

Even if this works perfectly in Django 3, I'm concerned that it may break backward compatibility with earlier versions (although once again, should have been caught by tests since we test against older versions of Django). Can you explain why that is not a concern?

Comment (tim-schilling) [reply django_cte/meta.py]: Unfortunately my answer is the one you're hoping for. I removed it since it was changing the `_alias` property that was used when generating the sql is no longer necessary. Since there's no need for a mutable property to keep track of the alias, there's no reason to change it elsewhere. I'm not familiar with the internals of the Django ORM, but I did test this against my application before creating the PR.

Do you know of a case in which relabeled_clone is called / used? If you can provide me with the general idea of it, I can make sure it's tested.

**Comment by ryanhiebert** [comment]
Interesting. The build failure was the connection to GitHub timing out on only one of them. Probably pushing a new commit with the same contents would fix the build.

File: django_cte/meta.py (hunk 1 above)
Comment (millerdev) [reply django_cte/meta.py]: I will see if I can come up with a test that breaks it. If I can't I'll consider it good and merge.

Comment (ryanhiebert) [reply django_cte/meta.py]: Thank you, @millerdev !

Comment (millerdev) [reply django_cte/meta.py]: I've been unable to write a failing test. In more detail (mostly for personal reference in case this ever becomes an issue later), I'm unable to find a scenario where `compiler.query.get_initial_alias()` returns the wrong result for a `CTEColumnRef`. It's hard to know all the possible values of `compiler.query`, but it looks like should always be a `CTEQuery`, given that `CTEColumnRef` is only created by `With.queryset()`:

https://github.com/dimagi/django-cte/blob/e999216ce2e0dd784060a8285bfe9470a867f683/django_cte/cte.py#L105-L111

Comment (millerdev) [reply django_cte/meta.py]: I ran some more tests and found a [failure](https://github.com/dimagi/commcare-hq/blob/0d0f6c238d64d52ceab07cf1c3ab82742a4e429e/corehq/apps/locations/tests/test_location_utils.py#L42)

```
======================================================================
FAIL: corehq.apps.locations.tests.test_location_utils:TestGetLocationsAndChildren.test_get_locations_and_children
----------------------------------------------------------------------
Traceback (most recent call last):
  File ".../corehq/apps/locations/tests/test_location_utils.py", line 49, in test_get_locations_and_children
    ['Middlesex', 'Cambridge', 'Somerville', 'Suffolk', 'Boston']
AssertionError: Element counts were not equal:
First has 2, Second has 1:  'Somerville'
```

I believe it is a result of this change.

Good SQL snippet (django-cte==1.1.4)
```sql
EXISTS (
    SELECT U0."id", U0."_cte_ordering" AS "_cte_ordering"
    FROM "xdups" U0
    WHERE (
		U0."id" = (cte."id")
		AND U0."_cte_ordering" = (cte."_cte_ordering")
	)
) AS "_exclude_dups"
```

Bad SQL snippet (this PR)
```sql
EXISTS (
    SELECT U0."id", U0."_cte_ordering" AS "_cte_ordering"
    FROM "xdups" U0
    WHERE (
    	U0."id" = (cte."id")
    	AND U0."_cte_ordering" = (U0."_cte_ordering")   -- NOTE wrong RHS alias
	)
) AS "_exclude_dups"
```

Dependency versions:
- Django 1.11.27
- Python 3.7.2

Comment (tim-schilling) [reply django_cte/meta.py]: This doesn't appear to be related to removing `relabed_clone`, but rather another symptom of the Django 3 changes that caused the original implementation to fail.

Comment (ryanhiebert) [reply django_cte/meta.py]: @millerdev : Does his response alleviate your concern for this particular pull request? If so, perhaps the right course of action would be to merge this, remove the "close # 10" from the description of this pull request, and create another issue for that?

Comment (tim-schilling) [reply django_cte/meta.py]: Actually the latest changes _should_ fix the issue @millerdev brought up. I was able to add a test for that specific case, then fix the problem.

Comment (ryanhiebert) [reply django_cte/meta.py]: Awesome! Thanks for clarifying that.

Comment (millerdev) [reply django_cte/meta.py]: I plan to run external tests against this again and make a decision based on that outcome. I'll try to do that soon.

**Comment by tim-schilling** [comment]
Sounds good. I tried setting up that project locally to run it, but gave up
as it was much larger than anticipated. Thank you for the quick responses.

**Comment by millerdev** [comment]
New failure, similar to the previous one.

```
ERROR: corehq.apps.locations.tests.test_location_fixtures:RelatedLocationFixturesTest.test_related_locations
----------------------------------------------------------------------
Traceback (most recent call last):
  File "lib/python3.7/site-packages/mock/mock.py", line 1305, in patched
    return func(*args, **keywargs)
  File "corehq/apps/locations/tests/test_location_fixtures.py", line 663, in test_related_locations
    flat=True
  File "lib/python3.7/site-packages/mock/mock.py", line 1305, in patched
    return func(*args, **keywargs)
  File "corehq/apps/locations/tests/test_location_fixtures.py", line 115, in _assert_fixture_matches_file
    fixture = ElementTree.tostring(call_fixture_generator(generator, self.user)[-1])
  File "corehq/ex-submodules/casexml/apps/phone/tests/utils.py", line 88, in call_fixture_generator
    return gen(restore_state)
  File "corehq/apps/locations/fixtures.py", line 121, in __call__
    locations_queryset = restore_user.get_locations_to_sync()
  File "lib/python3.7/site-packages/memoized.py", line 20, in _memoized
    cache[key] = value = fn(*args, **kwargs)
  File "corehq/ex-submodules/casexml/apps/phone/models.py", line 145, in get_locations_to_sync
    return get_location_fixture_queryset(self)
  File "corehq/apps/locations/fixtures.py", line 370, in get_location_fixture_queryset
    SQLLocation.objects.get_descendants(Q(domain=user.domain, id__in=user_location_ids))
  File "corehq/apps/locations/models.py", line 855, in from_locations
    related_locations = {loc_id for relation in relations for loc_id in relation}
  File "lib/python3.7/site-packages/django/db/models/query.py", line 250, in __iter__
    self._fetch_all()
  File "lib/python3.7/site-packages/django/db/models/query.py", line 1121, in _fetch_all
    self._result_cache = list(self._iterable_class(self))
  File "lib/python3.7/site-packages/django/db/models/query.py", line 122, in __iter__
    for row in compiler.results_iter(chunked_fetch=self.chunked_fetch):
  File "lib/python3.7/site-packages/django/db/models/sql/compiler.py", line 841, in results_iter
    results = self.execute_sql(MULTI, chunked_fetch=chunked_fetch)
  File "lib/python3.7/site-packages/django/db/models/sql/compiler.py", line 899, in execute_sql
    raise original_exception
  File "lib/python3.7/site-packages/django/db/models/sql/compiler.py", line 889, in execute_sql
    cursor.execute(sql, params)
  File "lib/python3.7/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "lib/python3.7/site-packages/django/db/utils.py", line 94, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "lib/python3.7/site-packages/django/utils/six.py", line 685, in reraise
    raise value.with_traceback(tb)
  File "lib/python3.7/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
django.db.utils.ProgrammingError: invalid reference to FROM-clause entry for table "cte"
LINE 1: ...RE (U0."id" = (V0."id") AND U0."_cte_ordering" = ("cte"."_ct...
                                                             ^
HINT:  Perhaps you meant to reference the table alias "v0".
```

Good SQL (django-cte==1.1.4)
```sql
WITH RECURSIVE cte AS ( ... )
SELECT V0."location_id" AS Col1
FROM "cte" V0
WHERE EXISTS (
    SELECT U0."id",
        U0."_cte_ordering" AS "_cte_ordering"
    FROM "xdups" U0
    WHERE (
        U0."id" = (V0."id")
        AND U0."_cte_ordering" = (V0."_cte_ordering")
    )
)

```

Bad SQL (bea2c9a8ec4de4fc6cb282e6182c33d9e007b2f2)
```sql
WITH RECURSIVE cte AS ( ... )
SELECT V0."location_id" AS Col1
FROM "cte" V0
WHERE EXISTS (
    SELECT U0."id", U0."_cte_ordering" AS "_cte_ordering"
    FROM "xdups" U0
    WHERE (
        U0."id" = (V0."id")
        AND U0."_cte_ordering" = ("cte"."_cte_ordering") -- NOTE wrong alias
    )
)
```

Dependency versions:

-    Django 1.11.27
-    Python 3.7.2

**Comment by tim-schilling** [comment]
Strange, I would have expected that to work as `"V0"` is an alias for `"cte"`. I'll take a look.

**Comment by tim-schilling** [comment]
@millerdev Could you include the whole SQL? It's a bit difficult to debug with just a partial set and without the ability to run commcare-hq's tests.

**Comment by tim-schilling** [comment]
@millerdev I was able to recreate the failure.

**Comment by tim-schilling** [comment]
@millerdev Alright, db45fe6 fixes the new test. I ended up having to re-add the private `_alias` property in case the `relabeled_clone` method was called.

File: django_cte/meta.py [hunk 2]
```diff
@@ -72,24 +72,29 @@ def relabeled_clone(self, relabels):
...
-        ref = self.copy()
-        ref._alias = query.get_initial_alias()
-        return ref
-
     def relabeled_clone(self, relabels):
-        if self._alias is not None and self._alias in relabels:
-            clone = self.copy()
+        if (
+            self.cte_name not in relabels
+            and self._alias not in relabels
```
[the remark this answers is not in the document]
Comment (tim-schilling) [reply django_cte/meta.py]: If I understood what happened correctly, `relabeled_clone` gets called when the query is used as a subquery. So if the wrapping query were then used as another subquery might trigger this again? It should be easy to check when I dive back in.

To answer your question, I don't know. My guess is that it could be a key in `relabels` simply because this method is based off of the `Col` expression's implementation.

**Comment by millerdev** [comment]
New failure, similar to the previous one.

```
ERROR: corehq.apps.locations.tests.test_location_fixtures:RelatedLocationFixturesTest.test_related_locations
----------------------------------------------------------------------
Traceback (most recent call last):
  File "lib/python3.7/site-packages/mock/mock.py", line 1305, in patched
    return func(*args, **keywargs)
  File "corehq/apps/locations/tests/test_location_fixtures.py", line 664, in test_related_locations
    flat=True
  File "lib/python3.7/site-packages/mock/mock.py", line 1305, in patched
    return func(*args, **keywargs)
  File "corehq/apps/locations/tests/test_location_fixtures.py", line 115, in _assert_fixture_matches_file
    fixture = ElementTree.tostring(call_fixture_generator(generator, self.user)[-1])
  File "corehq/ex-submodules/casexml/apps/phone/tests/utils.py", line 88, in call_fixture_generator
    return gen(restore_state)
  File "corehq/apps/locations/fixtures.py", line 121, in __call__
    locations_queryset = restore_user.get_locations_to_sync()
  File "lib/python3.7/site-packages/memoized.py", line 20, in _memoized
    cache[key] = value = fn(*args, **kwargs)
  File "corehq/ex-submodules/casexml/apps/phone/models.py", line 145, in get_locations_to_sync
    return get_location_fixture_queryset(self)
  File "corehq/apps/locations/fixtures.py", line 370, in get_location_fixture_queryset
    SQLLocation.objects.get_descendants(Q(domain=user.domain, id__in=user_location_ids))
  File "corehq/apps/locations/models.py", line 857, in from_locations
    return related_locations - {l.location_id for l in locations}
  File "lib/python3.7/site-packages/django/db/models/query.py", line 250, in __iter__
    self._fetch_all()
  File "lib/python3.7/site-packages/django/db/models/query.py", line 1121, in _fetch_all
    self._result_cache = list(self._iterable_class(self))
  File "lib/python3.7/site-packages/django/db/models/query.py", line 53, in __iter__
    results = compiler.execute_sql(chunked_fetch=self.chunked_fetch)
  File "lib/python3.7/site-packages/django/db/models/sql/compiler.py", line 899, in execute_sql
    raise original_exception
  File "lib/python3.7/site-packages/django/db/models/sql/compiler.py", line 889, in execute_sql
    cursor.execute(sql, params)
  File "lib/python3.7/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "lib/python3.7/site-packages/django/db/utils.py", line 94, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "lib/python3.7/site-packages/django/utils/six.py", line 685, in reraise
    raise value.with_traceback(tb)
  File "lib/python3.7/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
django.db.utils.ProgrammingError: missing FROM-clause entry for table "V0"
LINE 1: ...ducts", "cte"."supply_point_id", "cte"."user_id", "V0"."_cte...
```

Good SQL (django-cte==1.1.4)
```sql
WITH RECURSIVE cte ...
SELECT "cte"."id",
    ...
    cte."_cte_ordering" AS "_cte_ordering",
    EXISTS (
        SELECT U0."id", U0."_cte_ordering" AS "_cte_ordering"
        FROM "xdups" U0
        WHERE (
                U0."id" = (cte."id")
                AND U0."_cte_ordering" = (cte."_cte_ordering")
                )
        ) AS "_exclude_dups"
FROM "cte"
WHERE ...
```

Bad SQL (7ae8ae9162bc3ec26e6b9a914abe0c67bf9f434b)
```sql
WITH RECURSIVE cte ...
SELECT "cte"."id",
    ...
    "cte"."user_id",
    "V0"."_cte_ordering" AS "_cte_ordering",   -- NOTE wrong alias
    EXISTS (
        SELECT U0."id", U0."_cte_ordering" AS "_cte_ordering"
        FROM "xdups" U0
        WHERE (
                U0."id" = (cte."id")
                AND U0."_cte_ordering" = (cte."_cte_ordering")
                )
        ) AS "_exclude_dups"
FROM "cte"
WHERE ...
```

Dependency versions:

-    Django 1.11.27
-    Python 3.7.2

**Comment by tim-schilling** [comment]
Are these errors only on Django 1.11? If so, I'm leaning towards waiting
until April when support for that version drops.

**Comment by ryanhiebert** [comment]
Fwiw, I'm not in favor of waiting, I am in favor of dropping 1.11 support in anticiaption, since 3.0 is already out, so we're past the straddle-lts motivation.

**Comment by millerdev** [comment]
> Are these errors only on Django 1.11?

Not to my knowledge, that's just the version I've been able to easily test against. My guess is these same errors would occur wth Django 2.x as well.

We still depend on Django 1.11 (which is still supported, although admittedly that window is closing).

**Comment by tim-schilling** [comment]
If it's not too much to ask, could you run those tests with 2.2 and 3.0?

**Comment by millerdev** [comment]
I am unable to easily test against 2.2 or 3.0 yet (still working on that). However, I was able to test with 2.0 with the same results posted above.

File: django_cte/meta.py [hunk 3]
```diff
@@ -93,6 +93,15 @@ def relabeled_clone(self, relabels):
             clone._alias = relabels[self._alias]
         return clone
 
+    def resolve_expression(self, query=None, allow_joins=True, reuse=None,
```
Comment (millerdev) [inline django_cte/meta.py]: It would make a nicer overall diff (easier to review the entire PR) if this was moved above `relabeled_clone` where it was before.

Comment (tim-schilling) [reply django_cte/meta.py]: That's the plan. I was able to get commcare-hq to build and run the tests. I pushed this to run those tests locally. Once I can confirm it's working properly, I'll clean the PR up and let you know.

**Comment by tim-schilling** [comment]
@millerdev I _think_ this is good to go. I was able to reproduce the problem with commcare-hq, then reproduce that effect in `django-cte` as a test. Fix it. And confirmed that the tests in commcare-hq were fixed. I wasn't able to run all of them, but I did verify the package (locations) that the majority of the problems were coming from.

**Comment by millerdev** [comment]
@tim-schilling thanks for persisting despite the non-ideal testing workflow. I was able to run commcare-hq locations tests against it, and they all passed so I am satisfied. Do you want to get anything else in before I make a new release?

**Comment by tim-schilling** [comment]
Excellent. No, I think this is the last bit that we needed. Thanks for your work on the project!
